libres/lib/res_util/block_fs.cpp

```diff
@@ -539,62 +417,55 @@ static void block_fs_build_index(block_fs_type *block_fs,
-                            "datafile:%s - data will be lost, node_size:%d\n",
-                            file_node->node_offset, data_file.c_str(),
-                            file_node->node_size);
-                else
-                    fprintf(
-                        stderr,
+    for (;;) {
+        auto block = file_node_fread_alloc(block_fs->data_stream, &filename);
+        if (!block.has_value())
+            break;
+
+        if ((block->status == NODE_INVALID) ||
+            (block->status == NODE_WRITE_ACTIVE)) {
+            if (block->status == NODE_INVALID)
+                fprintf(stderr,
+                        "** Warning:: invalid node found at offset:%ld in "
+                        "datafile:%s - data will be lost, node_size:%d\n",
+                        block->node_offset, data_file.c_str(),
+                        block->node_size);
+            else
+                fprintf(stderr,
                         "** Warning:: file system was prematurely shut down "
                         "while writing node in %s/%ld - will be discarded.\n",
-                        data_file.c_str(), file_node->node_offset);
+                        data_file.c_str(), block->node_offset);
 
-                error_offset.push_back(file_node->node_offset);
-                file_node_free(file_node);
-                block_fs_fseek_valid_node(block_fs);
-            } else {
-                if (file_node_verify_end_tag(file_node,
-                                             block_fs->data_stream)) {
-                    block_fs_fseek_node_end(block_fs, file_node);
-                    block_fs_install_node(block_fs, file_node);
-                    if (file_node->status == NODE_IN_USE) {
-                        block_fs_insert_index_node(block_fs, filename,
-                                                   file_node);
-                    } else {
-                        util_abort("%s: node status flag:%d not recognized - "
-                                   "error in data file \n",
-                                   __func__, file_node->status);
-                    }
+            error_offset.push_back(block->node_offset);
+            block_fs_fseek_valid_node(block_fs);
+        } else {
+            if (file_node_verify_end_tag(*block, block_fs->data_stream)) {
+                block_fs_fseek_node_end(block_fs, *block);
+                if (block->status == NODE_IN_USE) {
+                    block_fs->index[filename] = *block;
                 } else {
-                    /*
+                    util_abort("%s: node status flag:%d not recognized - "
+                               "error in data file \n",
+                               __func__, block->status);
+                }
+            } else {
+                /*
              Could not find a valid END_TAG - indicating that
              the filesystem was shut down during the write of
              this node.  This node will NOT be added to the
```

Comment:
This comment should be reformatted to linecomments so as to not break indentation.